internal/authorizer/authorizer.go

```diff
@@ -56,90 +55,152 @@ func (a *Authorizer) Authorize(
 		if err != nil {
 			return types.DataResponseV1{},
-				&StatusCodeError{fmt.Errorf("failed to create a new authorization response: %w", err), http.StatusInternalServerError}
+				&StatusCodeError{fmt.Errorf("namespaced SAR failed: %w", err), http.StatusUnauthorized}
 		}
-	case CreateVerb:
-		// No namespace check needed as there won't be a query injection
-		res = minimalDataResponseV1(allowed)
-	default:
-		// Verb was already validated in handler; at this step, an unexpected verb is a bug
-		return types.DataResponseV1{}, &StatusCodeError{fmt.Errorf("unexpected verb: %s", verb), http.StatusInternalServerError}
+		level.Debug(a.logger).Log(
+			"msg", "namespace-scoped SubjectAccessReview",
+			"user", user, "groups", fmt.Sprintf("%s", groups),
+			"res", resource, "name", resourceName, "api", apiGroup,
+			"allowed", nsAllowed, "namespace", ns,
+		)
+
+		if nsAllowed {
+			allowed = append(allowed, ns)
+		}
+	}
+
+	if len(allowed) == 0 {
+		// all SARs were unsuccessful -> deny
+		return minimalDataResponseV1(false), nil
 	}
 
-	err = a.cache.Set(cacheKey, res)
+	// allow access for the namespaces where the SAR was successful
+	res, err := newDataResponseV1(allowed, a.matcher)
 	if err != nil {
 		return types.DataResponseV1{},
-			&StatusCodeError{fmt.Errorf("failed to store authorization response into cache: %w", err), http.StatusInternalServerError}
+			&StatusCodeError{fmt.Errorf("failed to create auth response: %w", err), http.StatusInternalServerError}
 	}
 
 	return res, nil
 }
 
-func generateCacheKey(
-	token, user string, groups []string,
-	verb, resource, resourceName, apiGroup string,
-) string {
-	userHash := hashUserinfo(token, user, groups)
-
-	return strings.Join([]string{
-		verb,
-		apiGroup, resourceName, resource,
-		userHash,
-	}, ",")
-}
+func (a *Authorizer) authorizeClusterWide(namespaces []string) (types.DataResponseV1, error) {
+	if a.matcher.IsEmpty() {
+		// user has cluster-wide access and does not need matcher -> allow
+		return minimalDataResponseV1(true), nil
+	}
+
+	// user has cluster-wide access but needs a matcher -> populate namespaces from API list
+	nsList, err := a.client.ListNamespaces()
+	if err != nil {
+		return types.DataResponseV1{}, &StatusCodeError{fmt.Errorf("failed to access api server: %w", err), http.StatusUnauthorized}
+	}
 
-func hashUserinfo(token, user string, groups []string) string {
-	hash := sha256.New()
-	hash.Write([]byte(token))
-	hash.Write([]byte(user))
+	if len(namespaces) == 0 {
+		// request was cluster-scoped, return matcher with all accessible namespaces
+		return newDataResponseV1(nsList, a.matcher)
+	}
+
+	nsMap := map[string]bool{}
+	for _, ns := range nsList {
+		nsMap[ns] = true
+	}
 
-	sort.Strings(groups)
-	for _, g := range groups {
-		hash.Write([]byte(g))
+	filtered := []string{}
+	for _, ns := range namespaces {
+		if nsMap[ns] {
+			filtered = append(filtered, ns)
+		}
 	}
```

Comment:
If this filtering to avoid giving permission to namespaces that are not on the API list?